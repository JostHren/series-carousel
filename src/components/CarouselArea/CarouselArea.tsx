import { selectSeriesQuery } from "../../store/series/seriesSlice";
import { TvMazeSeries } from "../../store/services/types";
import { useGetSeriesByNameQuery } from "../../store/services/series";
import { useSelector } from "react-redux";
import { getRandomFromArray } from "../../helpers";
import { useMemo } from "react";
import { HeroCarousel } from "../HeroCarousel/HeroCarousel";
import { SeriesCarousel } from "../SeriesCarousel/SeriesCarousel";
import {
  LOADING,
  ERROR,
  NO_SEARCH_QUERY,
  NO_RESULTS,
  NOT_POPULAR,
} from "./constants";

export const CarouselArea = (): JSX.Element => {
  const query = useSelector(selectSeriesQuery());
  const { error, isLoading, data, isFetching } = useGetSeriesByNameQuery(
    query ?? ""
  );

  const threeRandomSeries: TvMazeSeries[] = useMemo(
    () => getRandomFromArray(data ?? [], 3),
    [data]
  );

  const restOfSeries = useMemo(
    () =>
      data?.filter((show) => {
        const checkForSameSeriesId = (series: TvMazeSeries): boolean =>
          series.show.id === show.show.id;

        return !threeRandomSeries.some(checkForSameSeriesId);
      }),
    [data, threeRandomSeries]
  );

  if (isFetching || isLoading) return <HeroCarousel data={LOADING} />;
  if (error) return <HeroCarousel data={ERROR} />;
  if (!query || query === "") return <HeroCarousel data={NO_SEARCH_QUERY} />;
  if (!data || data.length === 0) return <HeroCarousel data={NO_RESULTS} />;

  return (
    <div className="flex flex-col sm:flex-row justify-between min-w-fit">
      <HeroCarousel data={threeRandomSeries} />
      <SeriesCarousel
        data={
          !restOfSeries || restOfSeries?.length === 0
            ? NOT_POPULAR
            : restOfSeries
        }
      />
    </div>
  );
};
