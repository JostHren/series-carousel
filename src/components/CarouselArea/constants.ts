export const NO_SEARCH_QUERY = [
  {
    show: {
      id: 0o0,
      image: { original: "./query.png" },
    },
  },
];

export const LOADING = [
  {
    show: {
      id: 0o0,
      image: { original: "./loading.png" },
    },
  },
];

export const ERROR = [
  {
    show: {
      id: 0o0,
      image: { original: "./error.png" },
    },
  },
];

export const NO_RESULTS = [
  {
    show: {
      id: 0o0,
      image: { original: "./no_results.png" },
    },
  },
];

export const NOT_POPULAR = [
  {
    show: {
      id: 0o0,
      image: { original: "./not_popular.png" },
      name: "Try something else",
    },
  },
];
