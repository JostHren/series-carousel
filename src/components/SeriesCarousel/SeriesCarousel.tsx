import { useCarouselNavigation, useCarouselNavigationProps } from "../../hooks";
import { CarouselProps, CarouselSlideProps } from "../HeroCarousel";

const SeriesCarouselLabel = (): JSX.Element => {
  return (
    <div className="absolute border-red-900 border-4 rounded-md bg-slate-900 p-2 text-red-800 text-lg font-extrabold right-5 top-3 rotate-3 z-50 shadow-md">
      Everything else
    </div>
  );
};

const SeriesCarouselNavigation = ({
  nextSlide,
  prevSlide,
}: Partial<useCarouselNavigationProps>): JSX.Element | null => {
  if (!nextSlide || !prevSlide) return null;

  return (
    <>
      <button
        className="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 top-0 left-0 right-0 z-10 h-10"
        onClick={() => prevSlide()}
      >
        <svg
          className="ml-2"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
          fill="#F6E05E"
          height={"50%"}
        >
          <path d="M3.29 15.71c.39.39 1.02.39 1.41 0l7.29-7.29 7.29 7.29a.996 .996 0 1 0 1.41-1.41l-7.98-8.01c-.2-.2-.45-.29-.71-.29s-.51.1-.71.29l-8 8a1 1 0 0 0 0 1.41Z"></path>
        </svg>
      </button>
      <button
        className="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 bottom-0 left-0 right-0 h-10 z-10"
        onClick={() => nextSlide()}
      >
        <svg
          className="ml-2"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
          fill="#F6E05E"
          height={"50%"}
        >
          <path d="M20.71 8.29a1 1 0 0 0-1.41 0l-7.29 7.29-7.29-7.29A.996 .996 0 1 0 3.31 9.7l8 8c.2.2.45.29.71.29s.51-.1.71-.29l8-8a1 1 0 0 0 0-1.41Z"></path>
        </svg>
      </button>
    </>
  );
};

const SeriesCarouselSlide = ({ slide }: CarouselSlideProps): JSX.Element => {
  return (
    <div className="flex mb-1 p-2 flex-row h-[18vh] sm:h-[22vh] w-full bg-slate-700 rounded-lg">
      <div className="basis-1/4 sm:basis-1/3 flex-grow-0 flex-shrink-0">
        <img
          src={slide.show?.image?.original ?? "./not_found.png"}
          alt={slide.show?.name}
          className="h-full m-auto"
        />
      </div>
      <div className="basis-3/4 sm:basis-2/3 flex-grow-0 flex-shrink-0 pl-4 text-yellow-400 truncate">
        <div className="text-lg font-extrabold truncate">
          {slide.show?.name}
        </div>
        <div>
          {slide.score
            ? `Rating: ${Math.round(slide.score * 100) / 100}`
            : null}
        </div>
        <div>{slide.show?.status ? `Status: ${slide.show.status}` : null}</div>
      </div>
    </div>
  );
};

export const SeriesCarousel = ({
  data,
  slideInterval = 3000,
  slidesPerPage = 2,
}: CarouselProps): JSX.Element => {
  const { isErrorSlide, currentIndex, prevSlide, nextSlide } =
    useCarouselNavigation({
      data,
      slideInterval,
      slidesPerPage,
    });

  return (
    <div className="rounded-lg m-2 overflow-hidden relative h-[37vh] sm:h-[90vh]">
      {!isErrorSlide && <SeriesCarouselLabel />}
      <SeriesCarouselNavigation prevSlide={prevSlide} nextSlide={nextSlide} />
      <div
        className="flex flex-col transition-transform duration-500 p-1"
        style={{
          transform: `translateY(-${currentIndex * 14.25}%)`,
        }}
      >
        {data.map((slide) => (
          <SeriesCarouselSlide slide={slide} key={slide.show?.id} />
        ))}
      </div>
    </div>
  );
};
