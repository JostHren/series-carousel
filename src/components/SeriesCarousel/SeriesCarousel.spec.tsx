import { render } from "@testing-library/react";

import { SeriesCarousel } from "./SeriesCarousel";
import { NOT_POPULAR } from "../CarouselArea/constants";

describe("SeriesCarousel component", () => {
  it("SeriesCarousel component to render", () => {
    const { container } = render(<SeriesCarousel data={NOT_POPULAR} />);

    expect(container).toMatchSnapshot();
  });
});
