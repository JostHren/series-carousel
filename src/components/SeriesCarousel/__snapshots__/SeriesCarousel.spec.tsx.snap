// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SeriesCarousel component > SeriesCarousel component to render 1`] = `
<div>
  <div
    class="rounded-lg sm:basis-2/6 m-2 overflow-hidden relative h-[calc(47dvh-6rem)] sm:h-[calc(99dvh-4rem)]"
  >
    <button
      class="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 top-0 left-0 right-0 z-10 h-10"
    >
      <svg
        class="ml-2"
        fill="#F6E05E"
        height="50%"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.29 15.71c.39.39 1.02.39 1.41 0l7.29-7.29 7.29 7.29a.996 .996 0 1 0 1.41-1.41l-7.98-8.01c-.2-.2-.45-.29-.71-.29s-.51.1-.71.29l-8 8a1 1 0 0 0 0 1.41Z"
        />
      </svg>
    </button>
    <button
      class="absolute bg-slate-800 bg-opacity-0 hover:bg-opacity-30 bottom-0 left-0 right-0 h-10 z-10"
    >
      <svg
        class="ml-2"
        fill="#F6E05E"
        height="50%"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.71 8.29a1 1 0 0 0-1.41 0l-7.29 7.29-7.29-7.29A.996 .996 0 1 0 3.31 9.7l8 8c.2.2.45.29.71.29s.51-.1.71-.29l8-8a1 1 0 0 0 0-1.41Z"
        />
      </svg>
    </button>
    <div
      class="flex flex-col transition-transform duration-500 p-1 pt-0"
      style="transform: translateY(-0%);"
    >
      <div
        class="
        h-[calc((47dvh-6.5rem)/1)] sm:h-[calc((99dvh-5rem)/2)] flex mb-1 p-2 pl-6 flex-row w-full bg-slate-700 rounded-lg"
      >
        <div
          class="basis-1/4 sm:basis-1/3 flex-grow-0 flex-shrink-0"
        >
          <img
            alt="Try something else"
            class="max-h-full aspect-auto m-auto"
            height="100%"
            src="./not_popular.png"
          />
        </div>
        <div
          class="basis-3/4 sm:basis-2/3 flex-grow-0 flex-shrink-0 pl-4 text-yellow-400 truncate"
        >
          <div
            class="text-lg font-extrabold truncate"
          >
            Try something else
          </div>
          <div />
          <div />
        </div>
      </div>
    </div>
  </div>
</div>
`;
